starlite/dto/abc.py

```diff
@@ -0,0 +1,301 @@
+
+            Add a new field called "new_field", that is a string, and required:
+            {"new_field": (str, ...)}
+
+            Add a new field called "new_field", that is a string, and not-required:
+            {"new_field": (str, "default")}
+
+            Add a new field called "new_field", that may be `None`:
+            {"new_field": (str | None, None)}
+        """
+        defined_fields: dict[str, FieldDefinition | NestedFieldDefinition] = {}
+        for field_definition in chain(cls.generate_field_definitions(model_type), cls.config.field_definitions):
+            if cls.should_exclude_field(field_definition):
+                continue
+
+            if field_mapping := cls.config.field_mapping.get(field_definition.field_name):
+                if isinstance(field_mapping, str):
+                    cls._reverse_field_mappings[field_mapping] = field_definition
+                    field_definition = copy(field_definition)  # noqa: PLW2901
+                    field_definition.field_name = field_mapping
+                else:
+                    cls._reverse_field_mappings[field_mapping.field_name] = field_definition
+                    field_definition = field_mapping  # noqa: PLW2901
+
+            if cls.detect_nested(field_definition):
+                nested_field_definition = cls.handle_nested(field_definition, nested_depth, recursive_depth)
+                if nested_field_definition is not None:
+                    defined_fields[field_definition.field_name] = nested_field_definition
+                continue
+
+            defined_fields[field_definition.field_name] = field_definition
+        return defined_fields
+
+    @classmethod
+    def handle_nested(
+        cls, field_definition: FieldDefinition, nested_depth: int, recursive_depth: int
+    ) -> NestedFieldDefinition | None:
+        if nested_depth == cls.config.max_nested_depth:
+            return None
+
+        args = get_args(field_definition.field_type)
+        origin = get_origin(field_definition.field_type)
+        nested = NestedFieldDefinition(
+            field_definition=field_definition,
+            origin=origin,
+            args=args,
+            nested_type=args[0] if args else field_definition.field_type,
+        )
+
+        if (is_recursive := nested.is_recursive(cls.model_type)) and recursive_depth == cls.config.max_nested_recursion:
+            return None
+
+        nested.nested_field_definitions = cls.parse_model(
+            nested.nested_type, nested_depth + 1, recursive_depth + is_recursive
+        )
+        return nested
+
+    @staticmethod
+    def get_model_type(item: type) -> Any:
+        """Get model type represented by the DTO.
+
+        Unwraps iterable annotation.
+
+        Args:
+            item: any type.
+
+        Returns:
+            The model type that is represented by the DTO.
+        """
+        if issubclass(get_origin(item) or item, Iterable):
```

Comment:
We have a predicate for this

Reply:
Which one are you referring to? `get_origin_or_inner_type()`, `is_class_and_subclass()`?

Just looking at this one:

```py
def get_origin_or_inner_type(annotation: Any) -> Any:
    """Get origin or unwrap it. Returns None for non-generic types.

    Args:
        annotation: A type annotation.

    Returns:
        Any type.
    """

    if origin := get_origin(annotation):
        return origin if origin not in (Annotated, Required, NotRequired) else get_args(annotation)[0]
    return None
```
do you think it should be recursive, i.e., return the same from both of these:

```python-console
>>> from starlite.utils.predicates import get_origin_or_inner_type
>>> from typing import List
>>> from typing_extensions import Annotated
>>> get_origin_or_inner_type(List[int])
<class 'list'>
>>> get_origin_or_inner_type(Annotated[List[int], None])
typing.List[int]
```

Reply:
is_iterable? No? If not, can you move this there?